lib/improver/tests/utilities/test_cube_units.py

```diff
@@ -34,38 +34,275 @@
+                self.percentile_cube, enforce=False)
+
+    def test_coord_datatype_fail(self):
+        """Test error is raised when enforce=False"""
+        self.percentile_cube.coord('percentile').points = (
+            self.percentile_cube.coord('percentile').points.astype(np.int32))
+        msg = "does not conform"
+        with self.assertRaisesRegex(ValueError, msg):
+            cube_units.enforce_units_and_dtypes(
+                self.percentile_cube, enforce=False)
+
+    def test_coordinates_correctly_identified(self):
+        """Test all coordinates in a heterogeneous cube list are identified and
+        corrected"""
+        self.percentile_cube.coord('percentile').points = (
+            self.percentile_cube.coord('percentile').points.astype(np.int32))
+        self.probability_cube.coord(
+            'air_temperature').convert_units('Fahrenheit')
+        result = cube_units.enforce_units_and_dtypes(
+            [self.percentile_cube, self.probability_cube])
+        self.assertEqual(result[0].coord('percentile').dtype, np.float32)
+        self.assertEqual(result[1].coord('air_temperature').units, 'K')
+
+    def test_subset_of_coordinates(self):
+        """Test function can enforce on a selected subset of coordinates and
+        leave all others unchanged"""
+        self.percentile_cube.coord('percentile').points = (
+            self.percentile_cube.coord('percentile').points.astype(np.int32))
+        self.percentile_cube.coord('time').convert_units(
+            'hours since 1970-01-01 00:00:00')
+        self.probability_cube.coord(
+            'air_temperature').convert_units('Fahrenheit')
+        self.probability_cube.coord('forecast_period').convert_units('h')
+
+        result = cube_units.enforce_units_and_dtypes(
+            [self.percentile_cube, self.probability_cube],
+            coords=["time", "forecast_period"])
+        self.assertEqual(result[0].coord('percentile').dtype, np.int32)
+        self.assertEqual(
+            result[0].coord('time').units, 'seconds since 1970-01-01 00:00:00')
+        self.assertEqual(
+            result[1].coord('air_temperature').units, 'Fahrenheit')
+        self.assertEqual(result[1].coord('forecast_period').units, 's')
+
+    def test_quantity_available(self):
+        """Test error raised if the named quantity is not listed in the
+        dictionary standard"""
+        self.data_cube.rename("number_of_fish")
+        self.data_cube.units = "1"
+        msg = "Name 'number_of_fish' not defined in units.py"
+        with self.assertRaisesRegex(KeyError, msg):
+            cube_units.enforce_units_and_dtypes(self.data_cube)
+
+
+# limited item dictionary to point to in smaller tests
+TEST_DICT = {
+    "time": {
+        "unit": "seconds since 1970-01-01 00:00:00",
+        "dtype": np.int64},
+    "probability": {"unit": "1"},
+    "temperature": {"unit": "K"},
+    "rainfall": {"unit": "m s-1"},
+    "rate": {"unit": "m s-1"}
+}
```

Comment:
It's quite unusual to have a global variable set up in the middle like this. An alternative would be to set up a test class that the others can inherit, e.g.

```
class local_test_dict(IrisTest):
    def setUp():
        self.test_dict = {
            "time": {
                "unit": "seconds since 1970-01-01 00:00:00",
                "dtype": np.int64},
            "probability": {"unit": "1"},
            "temperature": {"unit": "K"},
            "rainfall": {"unit": "m s-1"},
            "rate": {"unit": "m s-1"}
        }

class Test__find_dict_key(local_test_dict):
    """Test method to find suitable substring keys in dictionary"""

     @staticmethod
    def setUp():
        """Redirect to dummy dictionary"""
        super().SetUp()
        cube_units.DEFAULT_UNITS = self.test_dict
...
```

Each test class where you want to use this dictionary can inherit ```local_test_dict``` and use the ```super().SetUp()``` to get hold of the dictionary defined therein.